gprofiler/profilers/java.py

```diff
@@ -273,11 +302,91 @@ def make_application_metadata(self, process: Process) -> Dict[str, Any]:
+            jvm_flags = [{"error": e.stderr}]
+        except Exception as e:
+            logger.exception("Failed to collect JVM flags", pid=process.pid)
+            jvm_flags = [{"error": type(e).__name__}]
+
+        metadata = {
+            "java_version": version,
+            "libjvm_elfid": libjvm_elfid,
+            "jvm_flags": jvm_flags,
+        }
 
         metadata.update(super().make_application_metadata(process))
         return metadata
 
+    def get_jvm_flags_serialized(self, process: Process) -> Optional[List[Dict]]:
+        jvm_flags = self.get_jvm_flags(process)
+        if jvm_flags is None:
+            return None
+
+        return [flag.to_dict() for flag in sorted(jvm_flags, key=lambda flag: flag.name)]
+
+    def get_jvm_flags(self, process: Process) -> Optional[Iterable[JvmFlag]]:
+        if self.java_collect_jvm_flags == JavaFlagCollectionOptions.NONE:
+            return None
+
+        jvm_raw_flags: Iterable[JvmFlag] = filter(self.filter_jvm_flag, self.get_jvm_flags_raw(process))
+
+        if self.java_collect_jvm_flags == JavaFlagCollectionOptions.ALL:
+            return jvm_raw_flags
+        elif self.java_collect_jvm_flags == JavaFlagCollectionOptions.DEFAULT:
+            return filter(self.default_collection_filter_jvm_flag, jvm_raw_flags)
+        else:
+            assert isinstance(self.java_collect_jvm_flags, list), f"Unrecognized value: {self.java_collect_jvm_flags}"
+            requested_flags = [flag for flag in jvm_raw_flags if flag.name in self.java_collect_jvm_flags]
+            not_found_flags = set(self.java_collect_jvm_flags) - {flag.name for flag in requested_flags}
+
+            # log if the set is not empty
+            if not_found_flags:
+                logger.warning("Flags not found in jvm flags:", not_found_flags=not_found_flags)
+
+            return requested_flags
+
+    @staticmethod
+    def filter_jvm_flag(flag: JvmFlag) -> bool:
+        """
+        Filter out flags that are:
+        1. Flags that are of ccstr or ccstrlist type (type=ccstr, type=ccstrlist) - we have problems parsing them correctly # noqa: E501
+        2. Flags that are manageable (kind=manageable) - they might change during execution
+        """
+        if flag.type in ["ccstr", "ccstrlist"]:
+            return False
+
+        if "manageable" in flag.kind:
+            return False
+
+        return True
+
+    @staticmethod
+    def default_collection_filter_jvm_flag(flag: JvmFlag) -> bool:
+        """
+        Filter out flags that are:
+        1. Default flags (origin=default), that are constant for the JDK version
+        2. Flags that are non-production (kind=notproduct), platform dependent (kind=pd), or in development (kind=develop) # noqa: E501
+        """
+        if flag.origin == "default":
+            return False
+
+        if set(flag.kind).intersection({"notproduct", "pd", "develop"}):
+            return False
+
+        return True
+
+    @functools.lru_cache(maxsize=1024)
```

Comment:
Nitpick: you can cache the outer function, i.e after filtering, no need to include ALL data in the cache, only the data that passed filtering